internal/wasmldr/src/loader/compiled/tls.rs

```diff
@@ -0,0 +1,340 @@
+
+    async fn read_vectored<'a>(&self, bufs: &mut [IoSliceMut<'a>]) -> Result<u64, Error> {
+        let (cap, tls) = &mut *self.lck.write().unwrap();
+
+        if tls.wants_read() {
+            tls.read_tls(cap.deref_mut())?;
+            tls.process_new_packets()?;
+        }
+
+        let n = match tls.reader().read_vectored(bufs) {
+            Err(e) if e.kind() == ErrorKind::UnexpectedEof => 0,
+            Err(e) => return Err(e.into()),
+            Ok(n) => n,
+        };
+
+        Ok(n as u64)
+    }
+
+    async fn read_vectored_at<'a>(
+        &self,
+        _bufs: &mut [IoSliceMut<'a>],
+        _offset: u64,
+    ) -> Result<u64, Error> {
+        Err(Error::badf())
+    }
+
+    async fn write_vectored<'a>(&self, bufs: &[IoSlice<'a>]) -> Result<u64, Error> {
+        let (cap, tls) = &mut *self.lck.write().unwrap();
+
+        let n = tls.writer().write_vectored(bufs)?;
+
+        while tls.wants_write() {
+            tls.write_tls(cap.deref_mut())?;
+        }
+
+        Ok(n as u64)
+    }
+
+    async fn write_vectored_at<'a>(
+        &self,
+        _bufs: &[IoSlice<'a>],
+        _offset: u64,
+    ) -> Result<u64, Error> {
+        Err(Error::badf())
+    }
+
+    async fn seek(&self, pos: std::io::SeekFrom) -> Result<u64, Error> {
+        self.any.seek(pos).await
+    }
+
+    async fn peek(&self, _buf: &mut [u8]) -> Result<u64, Error> {
+        Err(Error::badf())
+    }
+
+    async fn num_ready_bytes(&self) -> Result<u64, Error> {
+        self.any.num_ready_bytes().await
+    }
+
+    fn isatty(&self) -> bool {
+        self.any.isatty()
+    }
+
+    async fn readable(&self) -> Result<(), Error> {
+        self.any.readable().await
+    }
+
+    async fn writable(&self) -> Result<(), Error> {
+        self.any.writable().await
+    }
+}
+
+pub struct Listener {
+    cap: Forgotten<CapListener>,
+    any: AnyListener,
+    cfg: Arc<ServerConfig>,
+}
+
+impl Listener {
+    pub fn new(tcp: cap_std::net::TcpListener, cfg: Arc<ServerConfig>) -> Self {
+        let cap = unsafe { CapListener::from_raw_fd(tcp.as_raw_fd()) }.into();
+        let any = AnyListener::from_cap_std(tcp);
+        Self { cap, any, cfg }
+    }
+}
+
+impl From<Listener> for Box<dyn WasiFile> {
+    fn from(value: Listener) -> Self {
+        Box::new(value)
+    }
+}
+
+#[wiggle::async_trait]
+impl WasiFile for Listener {
+    fn as_any(&self) -> &dyn Any {
+        self.any.as_any()
+    }
+
+    async fn sock_accept(&mut self, fdflags: FdFlags) -> Result<Box<dyn WasiFile>, Error> {
+        // Accept the connection.
+        let (mut cap, ..) = self.cap.accept()?;
+
+        // Create a new TLS connection.
+        let mut tls = Connection::Server(
+            ServerConnection::new(self.cfg.clone())
+                .map_err(|e| Error::io().context(e))
+                .context("could not create new TLS connection")?,
+        );
+
+        // Perform handshake.
+        cap.set_nonblocking(true)?;
+        tls.complete_io(&mut cap)
```

Comment:
Didn't we want to handle this within read and write implementations?
I believe we could just use `complete_io` in those

Reply:
Done. However, I think there are subtle bugs here in nonblocking mode. We shall see...